Display a warning when dune builds via RPC

gridbugs opened the pull request:
When an instance of dune is already running, invoking `dune build` will still build the target, howwever other command line arguments are silently ignored by dune as the current RPC interface is too limited to express them. This may change over time, but currently it's unusual UX for `dune build` to behave differently depending on whether or not another instance of `dune build` is already running. To more accurately set user expectations, print a warning when dune is running with this reduced functionality.

nojb commented on test/blackbox-tests/test-cases/watching/watching-eager-concurrent-build-command.t:
```diff
@@ -24,12 +24,20 @@ Build the project once before starting the watch server so the watch server star
+  Warning: A running dune (pid: 124305) instance has locked the build
+  directory. If this is not the case, please delete "_build/.lock".
+  The requested targets will still be built however functionality will be
+  reduced and some command line arguments will be ignored.
```
A couple of questions:

- "If this is not the case, please delete ..." is this relevant ? if the build is being forwarded to a running instance, it doesn't make much sense to me to ask the user to delete the lock file.
- "functionality will be reduced" what does this refer to?

I suggest tweaking the message and making it a bit less scary:

> Warning: your build request is being forwarded to a running Dune instance (pid: 124305). Note that certain command line arguments may be ignored.

Leonidas-from-XIV commented on bin/build_cmd.ml:
```diff
@@ -217,6 +217,12 @@ let build =
+          ; Pp.text
+              "The requested targets will still be built however functionality will be \
+               reduced and some command line arguments will be ignored."
+          ];
```
Would it be possible to adjust this error message to only display the warning when there are arguments that are being ignored? This way it warns about arguments being ignored even if there are no arguments.

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/watching/watching-eager-concurrent-build-command.t:
```diff
@@ -24,12 +24,20 @@ Build the project once before starting the watch server so the watch server star
 acting as an RPC client while running concurrently with an RPC server.
   $ echo 'let () = print_endlin "Hello, World!"' > foo.ml
   $ dune build
+  Warning: A running dune (pid: 124305) instance has locked the build
```
Isn't this PID going to change every time the test runs? One would probably need to override the PID in the lock file to make it reproducible.

Leonidas-from-XIV reviewed (commented):
I like the idea of warning the user when flags get ignored, but I would suggest that this would only happen if there are flags that will get discarded.

gridbugs committed:
Replace [Out_channel] log file with [Stderr]

The only [out_channel] that was ever passed in was stderr. Removing the
[Out_channel] case allows us to more easily test [Log.File.t] for
equality.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Add [Common.Builder.equal]

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Extract default [Builder.t] from term

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Expose lock held information when taking global lock

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Display a warning when dune builds via RPC

When an instance of dune is already running, invoking `dune build` will
still build the target, howwever other command line arguments are
silently ignored by dune as the current RPC interface is too limited to
express them. This may change over time, but currently it's unusual UX
for `dune build` to behave differently depending on whether or not
another instance of `dune build` is already running. To more accurately
set user expectations, print a warning when dune is running with this
reduced functionality.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

Leonidas-from-XIV commented on bin/build_cmd.ml:
```diff
@@ -215,6 +215,16 @@ let build =
+          User_warning.emit
+            [ Pp.textf
+                "Your build request is being forwarded to a running Dune instance%s. \
+                 Note that certain command line arguments may be ignored."
```
Since this warning is now only emitted when arguments have been ignored, I think it should state that command line arguments have in fact been ignored.

gridbugs replied:
Not all arguments are ignored though. The `Builder` module is perhaps poorly named, as `Builder.term` takes care of parsing arguments like `--verbose` and `--display` which still take effect when dune is behaving as an rpc client. I'll still think about how to reword this message to be more helpful though.

gridbugs replied:
I think the nuance is that we don't explicitly ignore arguments. Some arguments (e.g. `--verbose`, `--display`) change some parts of dune's global mutable state in ways that affect the output of the current instance of dune. These arguments are not ignored by `dune build` when it is acting as an rpc client; passing them will still cause dune to be more verbose or display output differently. Other arguments affect the behaviour of the build engine, and these are ignored by merit of the fact that the build engine is not run by dune when it's acting as an rpc client, and so incidentally these arguments are ignored.

Leonidas-from-XIV replied:
Yes, but we only display if the Builder instances are unequal so we know for a fact that we ignored command line arguments, there is no "may" at this point?

gridbugs replied:
Here's a little example usage of dune built from this branch. These commands were run while another dune was running in watch mode in a separate terminal:
```
2 s@u3 ~/src/llama (main=)
+ λ dune build examples/echo_effect.exe
Success

3 s@u3 ~/src/llama (main=)
+ λ dune build examples/echo_effect.exe --verbose
Shared cache: enabled
Shared cache location: /home/s/.cache/dune/db
Workspace root: /home/s/src/llama
Auto-detected concurrency: 32
Warning: Your build request is being forwarded to a running Dune instance
(pid: 347998) so most command-line arguments will be ignored.
Success
```

The builder instance was different from its default value because of `--verbose` however that flag is not ignored. When `Common.init` was called it set some global state including how dune should display its output, and this happens regardless of whether the builder is equal to its default value.

Leonidas-from-XIV replied:
Ok, thanks for the clarification. I did not expect `--verbose` be considered for `Builder`.

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/watching/watching-eager-concurrent-build-command.t:
```diff
@@ -26,6 +27,13 @@ Demonstrate that we can run "dune build" while the watch server is running.
+  Warning: Your build request is being forwarded to a running Dune instance
+  (pid: PID). Note that certain command line arguments may be ignored.
+  Success
+
```
Can you add an additional test case that shows that no such warning is printed if no arguments are ignored?

gridbugs replied:
Does the command on line 26 of this file satisfy that?

Leonidas-from-XIV commented on test/expect-tests/dune_rpc_e2e/dune_rpc_e2e.ml:
```diff
@@ -9,7 +9,7 @@ module Session = Csexp_rpc.Session
 (* enable to debug process stdout/stderr *)
 let debug = false
-let () = if debug then Dune_util.Log.init ~file:(Out_channel stderr) ()
+let () = if debug then Dune_util.Log.init ~file:Stderr ()
```
Curious that the only case where `Out_channel` is used was this but it does simplify the code as it is impossible to compare equality of channels so that's neat.

Leonidas-from-XIV reviewed (commented):
It's a bit sad all these very verbose `equal` functions need to be added but I understand the idea.

Overall I think it needs a small wording change and an additional test and then it is good to go.
